### PR TITLE
Add AddRaw method which does not follow symlinks

### PR DESCRIPTION
**Pull request by Code0x58**
This is a minor† change for #199 which on the surface adds a `AddRaw` method to the public interface which does not resolve symlinks before starting a watch.
 - [x] rewrite existing `Add` methods to be called `AddRaw` and not follow symlinks
 - [x] reimplement a generic `Add` method which uses `filepath.EvalSymlinks(name)` before calling `AddRaw`
 - [x] ~~consider adding atomic symlink update method~~ (separate issue)
 - [x] document

A follow on from this would be to offer a high level interface as mentioned on my initial comment for the issue - this can be tracked in a separate issue+PR.

† this does change the behaviour of `Add` on windows as it didn't follow symlinks (which do exist) before, but now it does

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -43,6 +43,9 @@ See [example_test.go](https://github.com/fsnotify/fsnotify/blob/master/example_t
 
 ## FAQ
 
+**Are symlinks followed?**
+As of v2.0.0 symlinks are no longer followed. You will have to explicitly get the underlying path with [`filepath.EvalSymlinks(path)`](https://golang.org/pkg/path/filepath/#EvalSymlinks) before passing it to the watcher if you want this behaviour.
```
Comment (nathany) [inline README.md]: Should we add an example (similar to example_test) of using this?

File: integration_test.go [hunk 2]
```diff
@@ -1012,6 +1011,71 @@ func TestFsnotifyClose(t *testing.T) {
...
+		t.Fatalf("Failed to create file1: %s", err)
+	}
+	defer f1.Close()
+	if _, err := os.Create(file2); err != nil {
+		t.Fatalf("Failed to create file2: %s", err)
+	}
+
+	// symlink works for Windows too
+	if err := os.Symlink(file1, link); err != nil {
+		t.Fatalf("Failed to create sylink: %s", err)
```
Comment (nathany) [inline integration_test.go]: typo: symlink

File: integration_test.go [hunk 3]
```diff
@@ -1012,6 +1011,71 @@ func TestFsnotifyClose(t *testing.T) {
...
+	select {
+	case event := <-w.Events:
+		t.Fatalf("Event from watcher: %v", event)
+	case err := <-w.Errors:
+		t.Fatalf("Error from watcher: %v", err)
+	case <-time.After(50 * time.Millisecond):
+	}
+
+	// ~atomic link change - event
+	tmpLink := filepath.Join(testDir, "tml-link")
```
Comment (nathany) [inline integration_test.go]: did you mean tmp-link?

File: integration_test.go [hunk 4]
```diff
@@ -1012,6 +1011,71 @@ func TestFsnotifyClose(t *testing.T) {
...
+	case _ = <-w.Events:
+	case err := <-w.Errors:
+		t.Fatalf("Error from watcher: %v", err)
+	case <-time.After(50 * time.Millisecond):
+		t.Fatalf("Took too long to wait for event")
+	}
+
+}
+
+/* XXX(obristow): these would return if implemeting the high level interfaced mentioned in https://github.com/fsnotify/fsnotify/issues/199#issuecomment-480277398
```
Comment (nathany) [inline integration_test.go]: Should we remove this test, or keep it and use EvalSymlinks?

**Comment by nathany** [comment]
This would also fix #227.

File: README.md (hunk 1 above)
Comment (Code0x58) [reply README.md]: I added a block to `example_test.go` which uses `filepath.EvalSymlinks(path)`, and linked to it here. I'm imagining having it in the standard example is better so people don't miss it.

File: integration_test.go (hunk 4 above)
Comment (Code0x58) [reply integration_test.go]: I pushed an amended version with these tests stripped, as I'm thinking we can dig them out of history if we want, and I don't think they'd be very meaningful if using `EvalSymlinks`.

File: README.md (hunk 1 above)
Comment (Code0x58) [reply README.md]: Now there's both `watcher.Add(name)` and `watcher.AddRaw(name)` in the example

File: integration_test.go (hunk 4 above)
Comment (Code0x58) [reply integration_test.go]: I've just reinstated these tests now that `watcher.Add(name)` is back to the symlink following behaviour

File: integration_test.go [hunk 5]
```diff
@@ -1007,11 +1007,74 @@ func TestFsnotifyClose(t *testing.T) {
...
+
+	err = w.AddRaw(link)
+	if err != nil {
+		t.Fatalf("Failed to add link: %s", err)
+	}
+
+	// change file 1 - no event
+	f1.Write([]byte("Hello"))
+	f1.Sync()
+	// XXX(Code0x58): doing a create here shows a CHMOD event on mac - is that an issue?
```
Comment (nathany) [inline integration_test.go]: 🤔

**Comment by Code0x58** [comment]
Would renaming to something like `AddWithoutDereferencing` be preferable?

**Comment by drewwells** [comment]
Can we get a release with this? Configmaps in k8s are mounted as configmaps, so fsnotify is of no use in those situations.

**Comment by ifilonenko** [comment]
What is the status on this? this would be great

File: fen.go [hunk 6]
```diff
@@ -26,8 +26,8 @@ func (w *Watcher) Close() error {
 	return nil
 }
 
-// Add starts watching the named file or directory (non-recursively).
-func (w *Watcher) Add(name string) error {
+// AddRaw starts watching the named file or directory (non-recursively). Symlinks are not implicitly resolved.
+func (w *Watcher) AddRaw(name string) error {
 	return nil
```
Comment (nathany) [inline fen.go:31]: this should probably do something

**Comment by nathany** [comment]
> Would renaming to something like `AddWithoutDereferencing` be preferable?

Perhaps.

Do people want to be able to add some paths to the same watcher that follow symlinks and some that don't? If not, maybe this could be a config option instead of a new API? E.g. `NewWatcher({FollowSymlinks: true})` or `watcher.FollowSymlinks(true)`?

File: fen.go (hunk 6 above)
Comment (Code0x58) [reply fen.go:31]: probably - I suppose it is covered by this existing [issue](https://github.com/fsnotify/fsnotify/issues/12)

**Comment by Code0x58** [comment]
one rationale for having separate adding methods in my mind is that it is explicit, so when reading a line of code, you know for sure what it will do at that point, without having to go back to the watcher's creation to see what it does, and similarly a change in one place won't have a significant behavioural change on other places. It may also be a little more convenient too, as you don't have to add `filepath.EvalSymlinks(path)` when following is off, or be stuck with following when following is on

**Review by niejian — APPROVED** [review]
LGTM

**Review by niejian — APPROVED** [review]
lgtm

**Comment by nathany** [comment]
Thanks. And apologies for the delay.

We need to sort out that CI issue #373.

**Comment by Code0x58** [comment]
Here's a PR for migrating away from Travis CI which shows green for master with go 1.11+ #378. I rebased a copy of this branch onto that and the build is green too [[test results](https://github.com/Code0x58/fsnotify/actions/runs/1087105544)]

**Review by nathany — DISMISSED** [review]

File: README.md [hunk 7]
```diff
@@ -90,6 +98,10 @@ See [example_test.go](https://github.com/fsnotify/fsnotify/blob/master/example_t
 
 ## FAQ
 
+**Are symlinks resolved?**
+Symlinks are implitily resolved by [`filepath.EvalSymlinks(path)`](https://golang.org/pkg/path/filepath/#EvalSymlinks) when `watcher.Add(name)` is used. If that is not desired, you can use `watcher.AddRaw(name)` to not follow any symlinks before watching. See [example_test.go](https://github.com/fsnotify/fsnotify/blob/master/example_test.go).
```
Comment (nathany) [inline README.md:102]: implitily -> implicitly

**Comment by nathany** [comment]
@Code0x58 I sent you a GitHub invite. Feel free to merge this.

Would you be willing to also update the CHANGELOG, AUTHORS, and tag a new release?

**Comment by Code0x58** [comment]
Thanks @nathany, I squashed the PR to avoid adding 2 other author emails for myself. It seems I can't push to master so will need an additional approval following the rebase, and might not be able to tag (not sure). Here's another PR to update the CHANGELOG and AUTHORS #380

**Review by nathany — APPROVED** [review]

**Pull request merged**